### PR TITLE
[ROCm] Add MXFP8 training support for gfx950 (MI355X)

### PR DESCRIPTION
**Pull request by indianspeedster**
### Summary

Adds ROCm gfx950 (MI355X) support for MXFP8 MoE training. Provides Triton kernel
replacements using `tl.dot_scaled` which feeds E8M0 block scales directly into
gfx950's native scaled MFMA instructions.

All changes are gated behind `is_ROCM()` — no impact on CUDA/SM100 paths.

### Changes

**New file:**
- `torchao/prototype/moe_training/kernels/mxfp8/rocm_mxfp8_mm.py` — Two Triton
  kernel entry points:
  - `triton_mxfp8_grouped_mm` — Persistent grouped GEMM for MoE fwd and dgrad.
    Grid = `num_CUs * ctas_per_cu`, each CTA walks the expert list with a
    global tile counter. No data-dependent grid, no silent row-dropping.
  - `triton_mxfp8_wgrad` — Weight gradient grouped GEMM. Internally dispatches
    to a split-K variant when `E == 1` (single-group grid too small to saturate
    the device); otherwise the direct path writes bf16 output in-place.

**Modified files:**

*MoE training:*
- `torchao/prototype/moe_training/mxfp8_grouped_mm.py` — ROCm dispatch for
  fwd / dgrad / wgrad, gated by `_ROCM_MXFP8_KERNELS_AVAILABLE`.
- `torchao/prototype/moe_training/config.py` — Add `mxfp8_dim1_cast_kernel_choice`
  field on `MXFP8TrainingOpConfig` so callers can pick `TRITON` on ROCm.
- `torchao/prototype/moe_training/tensor.py` — Thread the new config field
  through to the op.
- `torchao/prototype/moe_training/utils.py` — Skip `@nonstrict_trace` on
  ROCm; newer dynamo rejects the FakeTensor output of our custom-op kernels.
- `torchao/prototype/moe_training/kernels/mxfp8/__init__.py` — Export the
  two new ROCm kernel entry points.
- `torchao/prototype/moe_training/kernels/mxfp8/quant.py` — Register
  `torch_pad_token_groups` as a `torch.library.custom_op` with a
  `register_fake` implementation so it traces cleanly under torch.compile.

*MX formats (used by the shared-expert dense path):*
- `torchao/prototype/mx_formats/mx_tensor.py` — ROCm path uses `F.scaled_mm`
  with `ScalingType.BlockWise1x32` for the dense matmul.
- `torchao/prototype/mx_formats/mx_linear.py` — Select `TRITON` dim1 cast
  kernel on ROCm (CUDA dim1 kernel unavailable there).
- `torchao/prototype/mx_formats/kernels.py` — `ROW_TILE=64` autotune config
  and `n_rows` on the autotune key for better per-shape tuning of dim1 cast.

*Tests + benchmarks:*
- `test/prototype/moe_training/test_mxfp8_grouped_mm.py` — Drop
  `@skip_if_rocm` markers now that the ROCm path is correct.
- `benchmarks/prototype/moe_training/bench_2d_3d_grouped_gemm.py` — Extend
  the MXFP8 gate to MI350+; add the ROCm call path
  (`bench_mxfp8_grouped_mm_rocm`).
- `benchmarks/prototype/moe_training/benchmark_scaled_grouped_mm_dq.py` —
  Extend the MXFP8 gate to MI350+.

### Benchmarks

Env: 1× AMD MI355X (gfx950), ROCm 7.1, torch nightly `2.13.0.dev20260417+rocm7.1`, `triton-rocm==3.7.0`.

Detailed run info (raw logs + env): https://github.com/indianspeedster/ao/actions/runs/24588975436

**Full training step (`benchmark_scaled_grouped_mm_dq.py`, fwd + dgrad + wgrad + dim1 quant, `--alignment-size 32`):**

| M | N | K | G | bf16 fwd+bwd (μs) | mxfp8 fwd+bwd (μs) | fwd+bwd speedup | bf16 fwd (μs) | mxfp8 fwd (μs) | fwd speedup |
|---|---|---|---|-------------------|---------------------|------------------|----------------|------------------|--------------|
| 32768  | 8192 | 5120 | 1 | 6564  | 9227  | 0.71× | 1807 | 1810 | 1.00× |
| 32768  | 8192 | 5120 | 2 | 6927  | 9645  | 0.72× | 1882 | 1907 | 0.99× |
| 128000 | 8192 | 5120 | 1 | 25579 | 33931 | 0.75× | 7001 | 7052 | 0.99× |
| 128000 | 8192 | 5120 | 2 | 25714 | 35146 | 0.73× | 7047 | 7100 | 0.99× |
| 32768  | 2048 | 7168 | 4 | 2976  | 4208  | 0.71× | 807  | 929  | 0.87× |
| 32768  | 2048 | 7168 | 8 | 3411  | 4549  | 0.75× | 972  | 954  | 1.02× |
| 128000 | 2048 | 7168 | 4 | 9882  | 15501 | 0.64× | 2697 | 3625 | 0.74× |
| 128000 | 2048 | 7168 | 8 | 10318 | 15812 | 0.65× | 2737 | 3682 | 0.74× |

**Grouped GEMM fwd-only (`bench_2d_3d_grouped_gemm.py`):**

| E | M | N | K | bf16 (μs) | mxfp8 (μs) | speedup |
|---|---|---|---|-----------|-------------|---------|
| 1 | 16640 | 2048 | 2048 | 159 | 110 | 1.44× |
| 1 | 16640 | 2048 | 5120 | 326 | 285 | 1.14× |
| 1 | 16640 | 2048 | 8192 | 483 | 433 | 1.11× |
| 1 | 16640 | 5120 | 2048 | 317 | 257 | 1.23× |
| 1 | 16640 | 5120 | 5120 | 692 | 648 | 1.07× |
| 1 | 16640 | 5120 | 8192 | 1106 | 1019 | 1.09× |
| 1 | 16640 | 8192 | 2048 | 471 | 404 | 1.16× |
| 1 | 16640 | 8192 | 5120 | 1063 | 1005 | 1.06× |
| 1 | 16640 | 8192 | 8192 | 1646 | 1565 | 1.05× |
| 2 | 16640 | 2048 | 2048 | 159 | 114 | 1.39× |
| 2 | 16640 | 2048 | 5120 | 321 | 284 | 1.13× |
| 2 | 16640 | 2048 | 8192 | 473 | 435 | 1.09× |
| 2 | 16640 | 5120 | 2048 | 316 | 263 | 1.20× |
| 2 | 16640 | 5120 | 5120 | 635 | 663 | 0.96× |
| 2 | 16640 | 5120 | 8192 | 1013 | 1019 | 0.99× |
| 2 | 16640 | 8192 | 2048 | 448 | 421 | 1.06× |
| 2 | 16640 | 8192 | 5120 | 984 | 1026 | 0.96× |
| 2 | 16640 | 8192 | 8192 | 1574 | 1597 | 0.99× |
| 4 | 16640 | 2048 | 2048 | 193 | 114 | 1.69× |
| 4 | 16640 | 2048 | 5120 | 396 | 296 | 1.34× |
| 4 | 16640 | 2048 | 8192 | 548 | 441 | 1.25× |
| 4 | 16640 | 5120 | 2048 | 371 | 267 | 1.39× |
| 4 | 16640 | 5120 | 5120 | 700 | 682 | 1.03× |
| 4 | 16640 | 5120 | 8192 | 1036 | 1032 | 1.00× |
| 4 | 16640 | 8192 | 2048 | 549 | 420 | 1.31× |
| 4 | 16640 | 8192 | 5120 | 1088 | 1048 | 1.04× |
| 4 | 16640 | 8192 | 8192 | 1738 | 1618 | 1.07× |
| 8 | 16640 | 2048 | 2048 | 244 | 120 | 2.04× |
| 8 | 16640 | 2048 | 5120 | 559 | 316 | 1.77× |
| 8 | 16640 | 2048 | 8192 | 737 | 449 | 1.64× |
| 8 | 16640 | 5120 | 2048 | 429 | 271 | 1.59× |
| 8 | 16640 | 5120 | 5120 | 847 | 695 | 1.22× |
| 8 | 16640 | 5120 | 8192 | 1207 | 1060 | 1.14× |
| 8 | 16640 | 8192 | 2048 | 589 | 432 | 1.36× |
| 8 | 16640 | 8192 | 5120 | 1297 | 1098 | 1.18× |
| 8 | 16640 | 8192 | 8192 | 1719 | 1704 | 1.01× |

**Accuracy (`test_mxfp8_grouped_mm.py`):** 129 passed, 16 skipped.

### Note

This is the initial functional baseline; more optimizations are in progress
and will land soon.

### CODE REVIEW

File: torchao/prototype/moe_training/kernels/mxfp8/rocm_mxfp8_mm.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+            go_scale = tl.load(
+                GO_scales_ptr + n_offs[:, None] * GO_scales_stride_n + mb_offs[None, :] * GO_scales_stride_mb,
+                mask=n_mask[:, None], other=127,
+            )
+            ia_scale = tl.load(
+                IA_scales_ptr + k_offs[:, None] * IA_scales_stride_k + mb_offs[None, :] * IA_scales_stride_mb,
+                mask=k_mask[:, None], other=127,
+            )
+
+            acc = tl.dot_scaled(go_tile, go_scale, "e4m3", ia_tile, ia_scale, "e4m3", acc=acc, out_dtype=tl.float32)
```
Comment (danielvegamyhre) [inline torchao/prototype/moe_training/kernels/mxfp8/rocm_mxfp8_mm.py:219]: i see these triton gemm kernels have no benchmarks and from looking at them i don't see kernel design patterns that are needed for good performance (warp specialized persistent kernel, etc). have you measured the TFLOPs/sec achieved, vs theoretical max on the target hardware?

File: torchao/prototype/mx_formats/mx_linear.py [hunk 2]
```diff
@@ -62,7 +62,12 @@ def _to_mxfp8_then_scaled_mm(
     grad_elem_dtype = torch.float8_e4m3fn
     block_size = 32
     mxfp8_dim0_cast_kernel_choice = MXFP8Dim0CastKernelChoice.TRITON
-    mxfp8_dim1_cast_kernel_choice = MXFP8Dim1CastKernelChoice.CUDA
+    # On ROCm, CUDA dim1 kernels are not available; use Triton instead.
+    from torchao.utils import is_ROCM
+    mxfp8_dim1_cast_kernel_choice = (
```
Comment (vkuzo) [inline torchao/prototype/mx_formats/mx_linear.py]: don't automagically set it, make the user set this explicitly

motivation: https://docs.pytorch.org/docs/stable/community/design.html#principle-2-simple-over-easy

File: torchao/prototype/moe_training/kernels/mxfp8/rocm_mxfp8_mm.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+_rocm_mxfp8_available = is_ROCM() and _triton_kernels_available
+
+if _rocm_mxfp8_available:
+    import triton
+    import triton.language as tl
+
+    # ==================== Grouped GEMM (fwd / dgrad) ====================
+
+    @triton.jit
+    def _mxfp8_grouped_mm_kernel(
```
Comment (vkuzo) [inline torchao/prototype/moe_training/kernels/mxfp8/rocm_mxfp8_mm.py:38]: can we make sure all these kernels have numerical tests that pass - include that in that PR test plan

File: torchao/prototype/moe_training/kernels/mxfp8/rocm_mxfp8_mm.py (hunk 1 above)
Comment (vkuzo) [reply torchao/prototype/moe_training/kernels/mxfp8/rocm_mxfp8_mm.py]: +1, every kernel here should have performance benchmarks

**Review by vkuzo — CHANGES_REQUESTED** [review]
high level I agree with the direction of workaround until core support is better for mxfp8 on AMD

requesting changes for tests + benchmarks

File: torchao/prototype/mx_formats/mx_linear.py (hunk 2 above)
Comment (vkuzo) [reply torchao/prototype/mx_formats/mx_linear.py]: i think a better way would be:
1. make this configurable from the top level config (instead of hardcoding here)
2. keep the default as what is best for CUDA
3. make it easy for the user to choose the right value for ROCm if they want to (via docs, util function, etc)

**Comment by indianspeedster** [comment]
Thanks @danielvegamyhre, @vkuzo.

  Addressed the review:
  - Correctness: K/scale-tail masking, m_mask bounded by group_end + global M. 129 passed, 16 skipped (all SQNR thresholds met).
  - Kernel design: grouped MM is now a persistent kernel (no silent row-dropping, torch.compile-clean). Dense path switched to F.scaled_mm(BlockWise1x32). Wgrad default tile retuned.
  - Config: dropped the is_ROCM() auto-switch in mx_linear.py; mxfp8_dim1_cast_kernel_choice is now explicit (CUDA default).
  - Benchmarks: extended bench_2d_3d_grouped_gemm.py to run on MI350+.
  Accuracy + benchmarks from a gfx950 runner:
  https://github.com/indianspeedster/ao/actions/runs/24548974494

  This is the functional baseline — More optimizations are in progress and will land soon.

**Comment by danielvegamyhre** [comment]
@indianspeedster i see you added benchmark scripts, thanks - can you share the results in the PR description?

**Comment by indianspeedster** [comment]
@danielvegamyhre  Added Benchmark data. 
All the Accuracy and benchmark logs can be found here : https://github.com/indianspeedster/ao/actions/runs/24588975436

**Comment by vkuzo** [comment]
@indianspeedster thanks for sharing the benchmarks.  If I understand correctly, there is no e2e speedup and the kernels are not achieving good performance yet, right?

If I got that correctly, might be worth working on the kernels first and then checking the e2e integration in once the kernels are faster and there are actual e2e speedups.

**Comment by danielvegamyhre** [comment]
Yeah, to add to what @vkuzo said, IMO i don't really want to check in a bunch of new kernels + code that in total produce a substantial net slowdown vs bf16. I would encourage you to focus on one kernel at a time, and submit each one for review after optimizing it to produce some meaningful speedup. 

In particular, I notice the ROCM mxfp8 grouped gemm kernel itself seems very weak, which makes it difficult or impossible to achieve any net speedup with dynamic quantization, regardless of how fast the quantization kernels are. I would prioritize that first, and once the mxfp8 grouped GEMM produces 1.8x+ vs bf16 across a variety of realistic shapes, then move on to the quantization kernels.
